paddle/framework/net_design.md

```diff
@@ -0,0 +1,159 @@
+# Network Design
+
+`Network` is the container and controller of a set of operators in a network, 
+users can build a real network from a `NetDef` in protobuf message 
+and use `Network.Run()` to run all the operators in the network.
+
+The `Network` will
+
+- manage all the operators contained in the network.
+- not own any `Variable`.
+
+# API
+
+## NetworkBase
+To make the `Network` extendable, a base class is defined like this
+
+```c++
+// The minimum a network should be implemented.
+class NetworkBase {
+ public:
+  // `def` is a proto message that describe the structure of a network.
+  NetworkBase(const NetDef& def);
+
+  // run all the operators and return success(true) or not, all the
+  // variables are located in `scope`.
+  virtual bool Run(Scope* scope) = 0;
+
+ protected:
+  // keys of the input variables feed into the network.
+  std::vector<string> inputs_;
+  // keys of the corresponding output variables the network will mutate.
+  std::vector<string> outputs_;
+};
+```
+
+All network implementations should build networks from  a protobuf message which 
+describes the structure of a real network; `Run` method should be implemented by 
+all implementations to offer a universal method to forward or backward compute a network.
+
+A method of factory pattern can be defined like
+
+```c++
+std::unique<NetworkBase* CreateNet(const NetDef& def) {
+  switch (def.model_type()) {
+    case NN:
+      return new Network(def);
+    case Recursive:
+      return new RecursiveNet(def);
+    case Recurrent:
+      return new RecurrentNet(def);
+  }
+  return new Network(def);
+}
+```
+
+Network is designed as the container of operators, to make it more extendable,
+we decompling it from the related variable resources. 
+A `scope` is provided to `Run` so that the network structure can be reused 
+in different scopes.
+
+Finally, `NetworkBase` can be used as followed
+
+```c++
+Scope default_scope;
+auto net = CreateNet(def);
+net.Run(&default_scope);
+```
+
+
+## A Simple Network Implemention
+
+A very basic implemention is as followed, all it does is simply to run every operators in sequence.
+
+```c++
+class ScratchNet final : public NetworkBase {
+ public:
+  // Create a network describe by `def`.  NetDef is the definition of a network.
+  ScratchNet(const NetDef &def);
+
+  // Run all the operators with the `scope`, if no scope is provided, default
+  // scope will be used instead.
+  virtual bool Run(Scope *scope = nullptr) override;
+
+ protected:
+  // Create operators accordding to `def`.
+  bool CreateNet(const NetDef &def);
+
+  // Add a operator which is identified as `type` and has attributes described
+  // in `attrs`, the `inputs` are the keys of readonly input variables,
+  // `outputs` are keys of mutable output variables.
+  bool AddOp(const std::string &type, const std::vector<string> &inputs,
```

Comment:
Maybe addOp should be an interface of NetworkBase. Because there may be some instance will manipulate `NetworkBase` in CPP, such as adding gradient ops, adding optimization ops, etc.

If we want user passing a protobuf message contains backward ops, we need to write `AddGradientOps` in Python. It will let Paddle coupled with `Python` language, which we did not want to.